File: practica_2/codigos_optimizados/codigo_a_optimizar.py
# python_no_opt.py
# Versión: 1.1.0
# Autor: Salvador Estrada Pulido
# Fecha: 2026-03-09
# Descripción:
# Calcula las frecuencias de una lista de enteros, encuentra el modo
# y calcula la suma de los dígitos del modo.
#
# Entrada:
# Lista de números enteros
#
# Salida:
# Frecuencias
# Modo
# Suma de dígitos del modo
# Tiempo de ejecución (para benchmarking)
# MOD: v1.1.0 — optimización usando diccionario O(n) y medición de tiempo

import time
# inicio del benchmark
inicio = time.time()
# lista de números
numeros = [3, -1, 0, 5, -7, 0, 2, 3, 3, -1, 5, 5, 5]
# diccionario de frecuencias
frecuencias = {}
# recorrer lista
for n in numeros:
    # si ya existe incrementar
    if n in frecuencias:
        frecuencias[n] += 1
    # si no existe inicializar
    else:
        frecuencias[n] = 1
# variables para el modo
modo = None
max_cuenta = -1
# recorrer diccionario
for valor, cuenta in frecuencias.items():
    # comparar frecuencia
    if cuenta > max_cuenta:
        # actualizar valores
        max_cuenta = cuenta
        modo = valor
# copia del modo
x = modo
# convertir a positivo
if x < 0:
    x = -x
# suma de dígitos
suma_digitos = 0
while x > 0:
    suma_digitos = suma_digitos + (x % 10)
    x = x // 10
# fin del benchmark
fin = time.time()
tiempo = (fin - inicio) * 1000
# resultados
print("Frecuencias:", frecuencias)
print("Modo:", modo, "con cuenta:", max_cuenta)
print("Suma de dígitos del modo:", suma_digitos)
print("Tiempo de ejecución (ms):", tiempo)

File: practica_2/codigos_optimizados/codigo_a_optimizar.c
/*
Versión: 1.3.0
Autor: Saldaña Ramirez Cesar Augusto
Fecha: 2026-03-12

Descripción:
Encuentra todos los números primos desde 2 hasta N,
cuenta cuántos existen, suma su valor y clasifica
primos pares e impares.

Entrada:
N (entero)

Salida:
Número de primos
Suma de primos
Primos pares
Primos impares
Tiempo de ejecución
*/

/* 
MOD: v1.3.0 — Se arreglaron multiples problemas, entre ellos:
- Comentarios de cambio donde no hubo cambio.
- Codigo que realmente no funciona, se copio uno incorrecto y no 
se verifico antes de subirse al repositorio.
- Se eliminaron 2 if en if(es_primo) para solo agregar uno al inicio.
*/

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
int main() {
    clock_t inicio = clock();
    int N = 1000;
    int count_primos = 0;
    long long suma_primos = 0;
    int primos_pares = 0;
    int primos_impares = 0;
    if (N >= 2) { // Se agrego este if al inicio para descartar el 2.
        count_primos++;
        suma_primos += 2;
        primos_pares = 1;
    }
    for (int m = 3; m <= N; m += 2) {
        int es_primo = 1;
        int d = 3; // Aqui estaba el comentario mal puesto.
        while (d * d <= m) {
            if (m % d == 0) {
                es_primo = 0;
                break;
            }
            d += 2; // Aqui debia ir el comentario de la v1.2.0 y v1.2.1
        }
    if (es_primo) {
        count_primos++;
        suma_primos += m;
        primos_impares++;
        } // Se eliminarion 2 if
    }
    clock_t fin = clock();
    double tiempo = (double)(fin - inicio) / CLOCKS_PER_SEC * 1000;
    printf("Primos encontrados: %d\n", count_primos);
    printf("Suma de primos: %lld\n", suma_primos);
    printf("Primos pares: %d\n", primos_pares);
    printf("Primos impares: %d\n", primos_impares);
    printf("Tiempo de ejecucion (ms): %.3f\n", tiempo);
    return 0;
}

/*
El codigo corre en 0.000ms ya que corre aproximadamente en 0.02ms lo que 
es mas veloz que un tick (En lo que se mide el tiempo en esta version).
*/
